sdk/identity/azure-identity/azure/identity/_internal/shared_token_cache.py

```diff
@@ -0,0 +1,185 @@
+                continue
+        filtered_accounts.append(account)
+    return filtered_accounts
+
+
+class SharedTokenCacheBase(ABC):
+    def __init__(self, username=None, **kwargs):  # pylint:disable=unused-argument
+        # type: (Optional[str], **Any) -> None
+
+        self._authority = kwargs.pop("authority", KnownAuthorities.AZURE_PUBLIC_CLOUD)
+        self._username = username
+        self._tenant_id = kwargs.pop("tenant_id", None)
+
+        cache = kwargs.pop("_cache", None)  # for ease of testing
+
+        if not cache and sys.platform.startswith("win") and "LOCALAPPDATA" in os.environ:
+            from msal_extensions.token_cache import WindowsTokenCache
+
+            cache = WindowsTokenCache(cache_location=os.environ["LOCALAPPDATA"] + "/.IdentityService/msal.cache")
+
+            # prevent writing to the shared cache
+            # TODO: seperating deserializing access tokens from caching them would make this cleaner
+            cache.add = lambda *_, **__: None
+
+        if cache:
+            self._cache = cache
+            self._client = self._get_auth_client(cache=cache, **kwargs)  # type: Optional[AadClientBase]
+        else:
+            self._client = None
+
+    @abc.abstractmethod
+    def _get_auth_client(self, **kwargs):
+        # type: (**Any) -> AadClientBase
+        pass
+
+    def _get_cache_items_for_authority(self, credential_type):
+        # type: (TokenCache.CredentialType) -> List[CacheItem]
+        """yield cache items matching this credential's authority or one of its aliases"""
+
+        items = []
+        for item in self._cache.find(credential_type):
+            environment = item.get("environment")
+            if environment == self._authority or (
+                self._authority == KnownAuthorities.AZURE_PUBLIC_CLOUD and environment in _PUBLIC_CLOUD_ALIASES
+            ):
+                items.append(item)
+        return items
+
+    def _get_accounts_having_matching_refresh_tokens(self):
+        # type: () -> Iterable[CacheItem]
+        """returns an iterable of cached accounts which have a matching refresh token"""
+
+        refresh_tokens = self._get_cache_items_for_authority(TokenCache.CredentialType.REFRESH_TOKEN)
+        all_accounts = self._get_cache_items_for_authority(TokenCache.CredentialType.ACCOUNT)
+
+        accounts = {}
+        for refresh_token in refresh_tokens:
+            home_account_id = refresh_token.get("home_account_id")
+            if not home_account_id:
+                continue
+            for account in all_accounts:
+                # When the token has no family, msal.net falls back to matching client_id,
+                # which won't work for the shared cache because we don't know the IDs of
+                # all contributing apps. It should be unnecessary anyway because the
+                # apps should all belong to the family.
+                if home_account_id == account.get("home_account_id") and "family_id" in refresh_token:
```

Comment:
What is the case where the refresh token doesn't have `family_id` why does this need to be excluded?

Reply:
MSAL uses this internally to identify refresh tokens cached by first-party apps. I only make the check because MSAL does; I don't think it's necessary while only first-party apps are caching (i.e. for the foreseeable future?).